### ISSUE TITLE
Stricter type inference and enforcement?

### ISSUE THREAD

**Issue by deklanw** [gatsbyjs/gatsby#16310]
Is there a way to assume that every field seen in a frontmatter is non-nullable? It seems like that would be a reasonable option to have.

I solved it like this,

```js
exports.createSchemaCustomization = ({ actions }) => {
   const { createTypes } = actions;
   const typeDefs = `
    type Frontmatter {
      title: String!
      date(formatString: String fromNow: Boolean difference: String locale: String): Date!
      description: String!
      tags: [String!]!
    }

    type MarkdownRemark implements Node {
      html: String!
      frontmatter: Frontmatter!
    }
  `;
   createTypes(typeDefs);
};
```

Which is fine.

But, it all works except for the `html` field which still ends up being nullable. Why? And, I'm not even sure why that is nullable in the first place. In what situation would the html be null?

Also, will this schema be enforced in any way? If one of my blog posts, for example, are missing a `date` field will the CLI yell at me? I'm using this with graphql codegen for convenience, but that would be nice too.

Thanks

**Comment by stefanprobst** [comment]
@deklanw 
* Gatsby v2.2 has introduced new schema-related APIs like the `createTypes` action, but many plugins are still relying on the `setFieldsOnGraphQLNodeType` API, which - while *not* officially deprecated - will just overwrite user-provided field-types. see [here](https://github.com/gatsbyjs/gatsby/blob/944363288d51d2cd0b845dc46f028122fba9d9a9/packages/gatsby-transformer-remark/src/extend-node-type.js#L503) for the `html` field
* slightly unrelated: you can put the `dateformat` directive on the `Frontmatter.date` field so you don't have to manually state all those input args: `date: Date @dateformat`

**Comment by deklanw** [comment]
@stefanprobst 

That explains the html thing. Thanks for the date tip. I'm still wondering why html is nullable to begin with, and whether these types get enforced via CLI complaining at me.

**Comment by stefanprobst** [comment]
yes would probably make sense to make it not nullable.
if you want you *can* enforce this yourself with the `createResolvers` API, which is run as the last step in schema generation (i.e. after `setFieldsOnGraphQLNodeType`):
```js
// gatsby-node.js
exports.createResolvers = ({ createResolvers }) => {
  createResolvers({
    MarkdownRemark: {
      html: {
        type: 'String!',
      }
    }
  })
}
```

GraphQL *will* complain when there are null values on `GraphQLNonNull` fields. You can check the error you'll see with something like this:
```js
const { graphql, buildSchema } = require('graphql')
graphql(
  buildSchema('type Query { html: String! }'),
  'query { html }'
).then(result => console.error(result.errors))
```

**Comment by RIP21** [comment]
Hi everyone. I'm with the author, many fields like `timeToRead` and `html` and mostly all others are optional and it drives me nuts a little :) 

Why? Because I'm full-blown on TypeScript and https://graphql-code-generator.com/ and it's indeed much easier to navigate all those deep trees of nodes, edges, JSON files, and other things, also it's nice to be sure that everything is nicely typed and "there" based on typings generated from the schema and `queries`, `fragments` etc.

Basically, the fact that 99% of the fields in `remark` types are optional whether they are not, makes it hell on earth for 100% typed Gatsby sites. Also types like `[String]` is even worse, not only an array part needs null check but also all elements inside while in real life it almost never they will be nullable. 

So, I will be super happy if the types that plugins are generating will become stricter.

**Comment by deklanw** [comment]
@RIP21 
Yes, that is my motivation for this post also.

**Comment by RIP21** [comment]
@stefanprobst Can we get this tagged properly maybe? So it will be in backlog or so? Because it's kinda painful :)

**Comment by RIP21** [comment]
Not stale, still pain.

**Comment by RIP21** [comment]
This is still a problem.

**Comment by MorrisonCole** [comment]
New to Gatsby and have been running into this too - it's definitely painful for TS sites!

**Comment by RIP21** [comment]
It's not stale. Thanks.

**Comment by wardpeet** [comment]
We're moving official plugins to use the schema customization API like @stefanprobst mentioned. For now, you can do it yourself in gatsby-node.js. It's still going to be impossible to do this correctly 100% of the time. We could use inference and if a property is available everywhere mark it as non nullable.

cc @vladar

**Comment by vladar** [comment]
Yes, you can follow #20069 for schema customization updates. We will try to convert as much as possible to `non-null` as a part of this process.

**Comment by RIP21** [comment]
Not stale, dear bot.

**Comment by LekoArts** [comment]
I'll close this in favor of https://github.com/gatsbyjs/gatsby/issues/20069 as the original question was answered and nothing actionable is present. We'd be happy to receive PRs to update the plugins to use the stricter variant :)
